ColorTV Demo Swift/ColorTV Test App Swift/VideoViewController.swift

```diff
@@ -0,0 +1,128 @@
+    }
+    
+    @IBOutlet weak var loadingIndicator: UIActivityIndicatorView!
+    private var playPauseBtnRecognizer: UITapGestureRecognizer!
+    private var recommendationVC: COLORRecommendationViewController!
+    private var nextItemVC: COLORNextItemRecommendationViewController!
+    
+    var videoPlayer: AVPlayer? {
+        return videoView.playerLayer.player;
+    }
+    
+    override func loadView() {
+        super.loadView()
+        playPauseBtnRecognizer = UITapGestureRecognizer(target: self, action: #selector(playPauseButtonClicked))
+        playPauseBtnRecognizer.allowedPressTypes = [NSNumber(value: UIPressType.playPause.rawValue)]
+        view.backgroundColor = UIColor.black
+        loadingIndicator.startAnimating()
+        
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        view.addGestureRecognizer(playPauseBtnRecognizer)
+        videoPlayer?.addObserver(self, forKeyPath: "rate", options: [.new, .old], context: nil)
+        videoPlayer?.addObserver(self, forKeyPath: kPlaybackLikelyToKeepUpKeyPath, options: [.new, .old], context: nil)
+    }
+    
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        if (keyPath == "rate") {
+            if let rate = change?[.newKey] as? Float, rate == 0 {
+                if recommendationVC != nil, presentedViewController != recommendationVC {
+                    present(recommendationVC, animated: true, completion: nil)
+                }
+                
+                if let isLikely = videoPlayer?.currentItem?.isPlaybackLikelyToKeepUp, isLikely {
+                    loadingIndicator.stopAnimating()
+                } else {
+                    loadingIndicator.startAnimating()
+                }
+            }
+        } else if keyPath == kPlaybackLikelyToKeepUpKeyPath {
+            if let likelyToKeepUp = change?[.newKey] as? Bool {
+                likelyToKeepUp ? loadingIndicator.stopAnimating() : loadingIndicator.startAnimating()
+            }
+        }
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        self.videoPlayer?.removeObserver(self, forKeyPath: "rate")
+        self.videoPlayer?.removeObserver(self, forKeyPath: kPlaybackLikelyToKeepUpKeyPath)
+    }
+    
+    func loadVideo(with url: URL, identifier: String) {
+        var videoItem = AVPlayerItem(url: url)
+        
+        if let videoPlayer = videoPlayer {
+            videoPlayer.replaceCurrentItem(with: videoItem)
+        } else {
+            videoView.player = AVPlayer(playerItem: videoItem)
+        }
+        
+        COLORAdController.sharedAdController().contentRecommendationController(forPlacement: COLORAdFrameworkPlacementVideoStart, andVideoId: identifier) {
+            (vc: COLORRecommendationViewController?, error: Error?) in
```

Comment:
Early return pattern would make it easier to read, and flatten flow. We already have lot ofintendation because of dispatching.